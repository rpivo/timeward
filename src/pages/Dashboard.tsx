import React, { createContext, useReducer, useRef } from 'react';
import Alignment from '@components/Alignment';
import Graph from '@components/Graph';
import PieChart from '@components/PieChart';
import Section from '@components/Section';
import Tile from '@components/Tile';
import Timeline from '@components/Timeline';
import Timer from '@components/Timer';
import Timesheet from '@components/Timesheet';
import StyledDashboard from '@styles/pages/Dashboard.styled';

type Action = {
  type: 'button clicked';
  payload: number;
};
type Store = {
  label: string | undefined;
  seconds: number;
}[];

type DashboardContextType = {
  dispatch: React.Dispatch<Action>;
  store: Store;
}

export const DashboardContext = createContext({} as DashboardContextType);

const Dashboard: React.FC = (): JSX.Element => {

  const ref = useRef<HTMLInputElement>(null);

  const reducer = (state: Store, action: Action): Store => {
    switch (action.type) {
    case 'button clicked':
      return [
        ...state,
        {
          label: ref?.current?.value,
          seconds: action.payload,
        },
      ];
    default:
      return state;
    }
  };

  const dashboardStore = [
    {
      label: '',
      seconds: 0,
    },
  ];
  const [store, dispatch] = useReducer(reducer, dashboardStore);

  const handleBlur = (): void => console.log(ref?.current?.value);

  return (
    <StyledDashboard>
      <Section>
        <DashboardContext.Provider value={{ dispatch, store }}>
          <Tile>
            <Alignment horizontal>
              <input ref={ref} type='text' onBlur={(): void => handleBlur()}></input>
            </Alignment>
            <Alignment horizontal>
              <Timer />
            </Alignment>
            <Timesheet>
              <Timesheet.Record />
            </Timesheet>
          </Tile>
        </DashboardContext.Provider>
        <Tile>
          <Graph />
        </Tile>
      </Section>
      <PieChart />
      <Tile width='full'>
        <Timeline>
          <Timeline.Day day={0} />
          <Timeline.Day day={1} />
          <Timeline.Day day={2} />
          <Timeline.Day day={3} />
          <Timeline.Day day={4} />
          <Timeline.Day day={5} />
          <Timeline.Day day={6} />
        </Timeline>
      </Tile>
    </StyledDashboard >
  );
};

export default Dashboard;
