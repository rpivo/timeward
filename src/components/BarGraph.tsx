import React, { useContext } from 'react';
import { DashboardContext } from '@pages/Dashboard';
import StyledBarGraph from '@styles/components/BarGraph.styled';

const Bar: React.FC = (): JSX.Element => {
  const { store, totalSeconds } = useContext(DashboardContext);
  const heights: number[] = [];

  const getPreviousHeights = (index: number): number =>
    heights
      .slice(0, index)
      .reduce((acc, curr) => acc + curr, 0);

  return (
    <svg width='65'>
      {store.map((record, index, arr) => {
        const height = ~~((record.seconds / totalSeconds) * 100);
        heights.push(height);
        if (record.seconds) return (
          <rect
            id={`bar-monday-${index - 1}`}
            fill={record.color}
            key={index}
            height={arr[index + 1] ? `${height}%` : `${100 - getPreviousHeights(index)}%`}
            width='100%'
            x='0'
            y={`${getPreviousHeights(index)}%`}
          />
        );
        return null;
      }
      )}
    </svg>
  );
};

const BarGraph: React.FC = (): JSX.Element =>
  <StyledBarGraph>
    <Bar />
    <Bar />
    <Bar />
    <Bar />
    <Bar />
    <Bar />
    <Bar />
  </StyledBarGraph>;

export default BarGraph;
