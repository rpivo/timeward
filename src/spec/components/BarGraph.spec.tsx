import React from 'react';
import renderer from 'react-test-renderer';
import { DashboardContext } from '@pages/Dashboard';
import BarGraph from '@components/BarGraph';

describe('BarGraph', () => {
  it('should render correctly', () => {
    const dispatch = jest.fn();
    const store = [
      {
        label: 'work',
        seconds: 10,
      },
      {
        label: 'play',
        seconds: 20,
      },
    ];
    const totalSeconds = 0;
    const tree = renderer
      .create(
        <DashboardContext.Provider value={{ dispatch, store, totalSeconds }}>
          <BarGraph />
        </DashboardContext.Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
