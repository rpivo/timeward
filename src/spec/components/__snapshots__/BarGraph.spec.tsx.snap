// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BarGraph should render correctly 1`] = `
<div
  className="sc-AxirZ kEiYJF"
>
  <svg
    width="65"
  >
    <rect
      height="0%"
      id="bar-monday--1"
      width="100%"
      x="0"
      y="0%"
    />
    <rect
      height="100%"
      id="bar-monday-0"
      width="100%"
      x="0"
      y="0%"
    />
  </svg>
  <svg
    width="65"
  >
    <rect
      height="0%"
      id="bar-monday--1"
      width="100%"
      x="0"
      y="0%"
    />
    <rect
      height="100%"
      id="bar-monday-0"
      width="100%"
      x="0"
      y="0%"
    />
  </svg>
  <svg
    width="65"
  >
    <rect
      height="0%"
      id="bar-monday--1"
      width="100%"
      x="0"
      y="0%"
    />
    <rect
      height="100%"
      id="bar-monday-0"
      width="100%"
      x="0"
      y="0%"
    />
  </svg>
  <svg
    width="65"
  >
    <rect
      height="0%"
      id="bar-monday--1"
      width="100%"
      x="0"
      y="0%"
    />
    <rect
      height="100%"
      id="bar-monday-0"
      width="100%"
      x="0"
      y="0%"
    />
  </svg>
  <svg
    width="65"
  >
    <rect
      height="0%"
      id="bar-monday--1"
      width="100%"
      x="0"
      y="0%"
    />
    <rect
      height="100%"
      id="bar-monday-0"
      width="100%"
      x="0"
      y="0%"
    />
  </svg>
  <svg
    width="65"
  >
    <rect
      height="0%"
      id="bar-monday--1"
      width="100%"
      x="0"
      y="0%"
    />
    <rect
      height="100%"
      id="bar-monday-0"
      width="100%"
      x="0"
      y="0%"
    />
  </svg>
  <svg
    width="65"
  >
    <rect
      height="0%"
      id="bar-monday--1"
      width="100%"
      x="0"
      y="0%"
    />
    <rect
      height="100%"
      id="bar-monday-0"
      width="100%"
      x="0"
      y="0%"
    />
  </svg>
</div>
`;
